testing/sdk_metrics.py

```diff
@@ -24,6 +25,22 @@ def get_scheduler_metrics(service_name, timeout_seconds=15 * 60):
     return sdk_cmd.service_request("GET", service_name, "/v1/metrics").json()
 
 
+def get_failure_metrics(service_name: str) -> typing.Dict:
+    history = sdk_cmd.history_request().json()
+    service_history = [h for h in history["frameworks"] if h["name"] == service_name]
+    if not service_history:
+        return dict()
+
+    def collect():
```

Comment:
A couple of things:
* We assume a length of the `service_history` list. Should we assert if we find multiple services with the same name?
* Then, we each of the `frameworks` that are in `service_history` have a known structure. Would it not make sense to just call `get()` on the desired failure keys?
* Lastly, what about the other failure modes: `TASK_LOST`, `TASK_UNREACHABLE`, or `TASK_KILLED`?

Reply:
- Assert for only one service sounds good
- The only assumed structure currently is the `name` field. I'll add a `get` with default to empty then it behaves like the service is not found and does not crash. 
- I am not completely familiar with what kind of status is raised when. Maybe you can give an update on that.

Reply:
One of the `"frameworks"` looks as follows:
```json
{
"id": "850f4418-2fc0-429f-9b08-e8530e42d8f4-0001",
"name": "marathon",
"pid": "scheduler-e851e3a7-2c86-497a-970b-b1a36766b06f@10.0.7.148:15101",
"used_resources": {
"disk": 0,
"mem": 0,
"gpus": 0,
"cpus": 0
},
"offered_resources": {
"disk": 0,
"mem": 0,
"gpus": 0,
"cpus": 0
},
"capabilities": [
"TASK_KILLING_STATE",
"GPU_RESOURCES",
"PARTITION_AWARE",
"REGION_AWARE"
],
"hostname": "10.0.7.148",
"webui_url": "https://10.0.7.148:8443",
"active": true,
"connected": true,
"recovered": false,
"TASK_STAGING": 0,
"TASK_STARTING": 0,
"TASK_RUNNING": 0,
"TASK_KILLING": 0,
"TASK_FINISHED": 0,
"TASK_KILLED": 0,
"TASK_FAILED": 0,
"TASK_LOST": 0,
"TASK_ERROR": 0,
"TASK_UNREACHABLE": 0,
"slave_ids": []
}
```
I would then explicitly do the following (or using an explicit `for` loop with `yield` and `collect`):
```python
framework_history = service_history[0]
fields_of_interest = ["TASK_FAILED", "TASK_ERROR"]
return dict([f, framework_history.get(f, 0) for f in fields_of_interest])
```
To me this is easier to read, and I would be willing to wage that this is faster as we are not iterating over a large number of key-value pairs just to find the ones of interest, but using the fast lookups in the dict.

Reply:
An overview of the mesos states and their reasons can be found here: http://mesos.apache.org/documentation/latest/task-state-reasons/#reasons-for-terminal-status-updates

Reply:
I now have used this reference: https://github.com/apache/mesos/blob/73d1da0893bab8cb7c5d540f115c5106f3e2039e/include/mesos/mesos.proto#L2212 
to compile the following state list 
```
FATAL_TERMINAL_TASK_STATES = set(
    ["TASK_FAILED", "TASK_ERROR", "TASK_KILLED", "TASK_DROPPED", "TASK_GONE"]
)
```